pkg/controllers/sync/controller.go

```diff
@@ -1104,58 +1111,120 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
+		return worker.Result{RequeueAfter: &s.cascadingDeletionRecheckDelay}
+	}
+
+	// cascading-delete is enabled, delete the member objects managed by admiral
 	ftcLister := s.ftcManager.GetFederatedTypeConfigLister()
 	ftcs, err := ftcLister.List(labels.Everything())
 	if err != nil {
-		logger.Error(err, "failed to get ftc lister")
+		logger.Error(err, "Failed to get ftc lister")
 		return worker.StatusError
 	}
 
-	remainingByGVK := make(map[string]string, len(ftcs))
-	for _, ftc := range ftcs {
-		gvk := ftc.GetSourceTypeGVK().String()
-		resourceLister, hasSynced, exists := s.fedInformerManager.GetResourceLister(
-			ftc.GetSourceTypeGVK(),
-			cluster.Name,
-		)
-		if !exists {
-			remainingByGVK[gvk] = fmt.Sprintf("failed to get resource lister for %s", gvk)
-			continue
-		}
+	// concurrent operations for each ftc
+	wg := sync.WaitGroup{}
+	syncMap := sync.Map{}
+	for i := range ftcs {
+		wg.Add(1)
+		go func(ftc *fedcorev1a1.FederatedTypeConfig) {
+			defer wg.Done()
+			gvk := ftc.GetSourceTypeGVK().String()
 
-		// If cluster cache is synced, we check the store.
-		// Otherwise, we will have to issue a list request.
-		if hasSynced() {
-			objects, err := resourceLister.List(labels.Everything())
+			// get all managed cluster objects according to ftc type
+			clusterObjs, namespaces, err := s.listManagedClusterObjects(ctx, ftc, cluster.Name)
 			if err != nil {
-				remainingByGVK[gvk] = fmt.Sprintf("Unknown (failed to list from cluster lister: %v)", err)
-			} else if len(objects) > 0 {
-				remainingByGVK[gvk] = strconv.Itoa(len(objects))
+				logger.Error(err, "Failed to list clusterObjects")
+				syncMap.Store(gvk, fmt.Sprintf("failed to list cluster objects for %s", gvk))
+				return
 			}
-		} else {
-			client, exists := s.fedInformerManager.GetClusterDynamicClient(cluster.Name)
-			if !exists {
-				remainingByGVK[gvk] = "Unknown (cluster client does not exist)"
-				continue
+			if len(clusterObjs) == 0 {
+				return
 			}
 
-			objects, err := client.Resource(ftc.GetSourceTypeGVR()).Namespace(corev1.NamespaceAll).List(
-				ctx, metav1.ListOptions{
-					Limit: 1,
-					LabelSelector: labels.SelectorFromSet(labels.Set{
-						managedlabel.ManagedByKubeAdmiralLabelKey: managedlabel.ManagedByKubeAdmiralLabelValue,
-					}).String(),
-				},
-			)
-			if err == nil && len(objects.Items) > 0 {
-				remainingByGVK[gvk] = strconv.Itoa(len(objects.Items))
-			} else if err != nil && !meta.IsNoMatchError(err) && !apierrors.IsNotFound(err) {
-				remainingByGVK[gvk] = fmt.Sprintf("Unknown (failed to list from cluster: %v)", err)
+			client, err := s.getClusterClient(cluster.Name)
+			if err != nil {
+				logger.Error(err, "Failed to get cluster client")
+				syncMap.Store(gvk, "failed to get cluster client")
+				return
 			}
-		}
+
+			// update some cluster objects before deleting them
+			var updateFailedObjs []string
+			for i := range clusterObjs {
+				clusterObj := clusterObjs[i].DeepCopy()
+				keyedLogger := logger.WithValues(
```

Comment:
Let's use  `logging.InjectLoggerValues`

Reply:
The code here has been changed to use `logging.InjectLoggerValues`.